Add JSONSchema definitions for services & validate using CI

WizardCM opened the pull request:
### Description

This adds full JSONSchema definition files for both services.json and package.json, which can be validated using various tools, including an IDE. This includes user-friendly descriptions pulled from [Service Submission Guidelines](https://github.com/obsproject/obs-studio/wiki/Service-Submission-Guidelines).

![image](https://user-images.githubusercontent.com/941350/163961215-100d99a2-4226-4532-9eaa-aab2ba221384.png)

![image](https://user-images.githubusercontent.com/941350/163960917-0285ff32-8eed-4243-97a4-6cb760bd9f0f.png)

![image](https://user-images.githubusercontent.com/941350/163960968-ed2ef177-fadc-4780-896f-2171dfc26586.png)

![image](https://user-images.githubusercontent.com/941350/163961044-1fa6590e-8335-4b81-a9b6-a34f099eca0e.png)

![image](https://user-images.githubusercontent.com/941350/163961146-14cf7f40-0527-4a48-aa71-329162e9bb32.png)


Additionally, this adds a Python-based validation runner when users submit Pull Requests, informing them of any failing validations directly within the diff.

![image](https://user-images.githubusercontent.com/941350/163961341-fb376c18-b3a0-4ec8-b7a7-f4202bd5f09c.png)

![image](https://user-images.githubusercontent.com/941350/163961381-b2fd53e5-4e21-4156-b238-bedd97994dbd.png)

Unfortunately, I had to write a custom python script to perform complete validation, as none of the existing libraries provide line numbers by default - it's exclusively _paths_ within the JSON structure. Additionally, errors would be poorly formatted and generally hard to read. Instead I have to return a custom JSON result and write that to a file, to be read by an annotation GitHub Action.

By default, CI will only validate the services.json file. If needed, I can add the package.json to the CI run too.

Bonus: json_source_map was built incorrectly and only works on Python 3.9 and above, so I had to manually install that too.

If the services.json format is changed, the schema should be bumped too, which is why the file is v3.

### Motivation and Context

When new services are submitted, there isn't enough feedback to the submitter about what was done incorrectly.

### How Has This Been Tested?

Install Python 3.9 and the required dependencies.

```sh
sudo apt install python3.9-dev
python3.9 -m pip install jsonschema json_source_map
```

Then run the validation script.

```sh
python3.9 CI/check-jsonschema.py plugins/rtmp-services/data/services.json plugins/rtmp-services/data/schema/service-schema-v3.json
```

Note that it silently succeeds. Modify the services.json to be invalid, then run the script again.

Additionally, ran the CI script manually.

### Types of changes

 - New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

derrod commented on .github/workflows/services-json.yml:
```diff
@@ -0,0 +1,36 @@
+name: Services Validator
+
+on:
+  push:
```
The push rule is missing the path filter, I guess this was temporary for testing? Otherwise it should use the same path filter. Also since the annotation doesn't happen for non-PRs how is the failure logged?

WizardCM replied:
If any validation errors are found and it's not a PR, it'll be treated properly as an error in the 'Validate Service JSON Schema' step itself, as the script will return a status code of 1 via `sys.exit(1)`.

derrod replied:
Does the action support adding annotations to files in a commit rather than just leaving the errors in this log?

WizardCM replied:
I am actually unsure. It's likely. I'll test it and get back to you.

WizardCM replied:
I have confirmed that annotations do work on straight commits, so we could enable the Annotate Errors step on all runs.

![image](https://user-images.githubusercontent.com/941350/164885587-be2303e5-4448-4591-a268-c06007d43994.png)

PatTheMav replied:
Yeah I also think it might be better to have CI just check for errors but committers should run the validator locally before committing anyway, so a local check should then give the explicit error anyway.

derrod commented on .github/workflows/services-json.yml:
```diff
@@ -0,0 +1,36 @@
+  push:
+    paths-ignore:
+      - "cmake/**"
+    branches-ignore: [master]
```
Ignoring master branch is probably not desired. Since it's the only one we really use for active development.

derrod commented on .github/workflows/services-json.yml:
```diff
@@ -0,0 +1,36 @@
+    paths:
+      - "plugins/rtmp-services/data/services.json"
+      - "plugins/rtmp-services/data/package.json"
+    branches-ignore: [master]
```
The ignore filter is evaluated against the target branch, not source branch, so this is not doing what you probably intended.

derrod commented on .github/workflows/services-json.yml:
```diff
@@ -0,0 +1,36 @@
+        uses: actions/checkout@v3
+        with:
+          submodules: "recursive"
+      - name: Install & Configure Python
```
Nitpick: for better readability I personally prefer having a blank line between steps

derrod commented on .github/workflows/services-json.yml:
```diff
@@ -0,0 +1,36 @@
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          submodules: "recursive"
```
I don't think we need to check out submodules here?

derrod reviewed (commented):
Just a first pass with some minor things I noticed without checking the python stuff itself.

WizardCM commented on plugins/rtmp-services/data/schema/service-schema-v3.json:
```diff
@@ -0,0 +1,206 @@
+                            "profile": {
+                                "type": "string",
+                                "description": "H.264 Profile.",
+                                "minLength": 1
```
I did consider doing that - I'll add it.

RytoEX commented:
It kind of bothers me that a lot of the description strings use punctuation while not actually being complete sentences. At the moment, I'm not sure how best to address that concern though, since the easiest way to fix that is adding something like "This value is/represents/controls/etc." to the beginning of each sentence, and I understand that that may not be ideal in some cases.

Overall, I am in favor of the concept in this PR though. If I can't come up with any improvements to the language, then I can always defer those to a later time.

PatTheMav commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,45 @@
+import json
```
As I mentioned in the related PR (https://github.com/obsproject/obs-studio/pull/6360#discussion_r856933794) I'd prefer if this script uses a `main` method and also handles file access exceptions.

WizardCM commented:
Rebased to master and rewrote the Python script based on feedback from @derrod and @PatTheMav - 

* there's now a `main` function
* script parameters are individual files to be validated
* CI will now validate both services.json and package.json
* CI will annotate regardless of whether it's a PR or a commit.

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def prep(filename):
+    global json_data
+    global json_string
+    global schema
```
Rather than using globals, why not just assing them here and pass them to the next function (since they're not used anywhere else)?

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,90 @@
+    schema_filename = json_data.get('$schema')
+    if not schema_filename:
+        print('File has no schema:', filename)
+        return 1
```
The return value of this function is never used.

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def validate(filename):
+    global errors
```
`errors` doesn't need to be declared global if you're not reassigning it.

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,90 @@
+    with open('validation_errors.json', 'w') as outfile:
+        json.dump(errors, outfile)
+
+    if len(errors) > 0:
```
Personal taste thing, but `if errors:` works fine as well as `errors` is only truthy if it is non-empty.

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,90 @@
+    for filename in inputArgs[1:]:
+        prep(filename)
+
+    with open('validation_errors.json', 'w') as outfile:
```
Missing exception handling when writing the output file.

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,90 @@
+        print("JSON path required.")
+        return 1
+
+    for filename in inputArgs[1:]:
```
Instead of declaring `inputArgs` just use `sys.argv` again and remove the unnecessary assignment at the top

derrod reviewed (changes requested):
I've had a couple issues with the python script speciifcally. Just a few oddities probably as a result from refactoring.

PatTheMav commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,90 @@
+
+    for e in sorted(cls.iter_errors(json_data), key=str):
+        print(e)
+        print (f'    in {filename}')
```
Extraneous whitespace between function name and parameters.

WizardCM commented:
All feedback actioned. Ready for re-review.

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,88 @@
+            json_data = json.loads(json_string)
+    except OSError as e:
+        print(f'Failed to load file "{filename}": {e}')
+        return 1
```
This still returns >:(

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,88 @@
+        print(f'Failed to load schema file "{schema_file}": {e}')
+        return
+
+    return validate(filename, json_data, json_string, schema)
```
You can just remove the `return` here.

derrod commented on CI/check-jsonschema.py:
```diff
@@ -0,0 +1,88 @@
+
+
+def prep(filename):
+    json_data = {}
+    json_string = ''
+    schema = {}
```
You don't need to assign them here.

derrod commented on plugins/rtmp-services/data/schema/service-schema-v3.json:
```diff
@@ -0,0 +1,211 @@
+                                "description": "OBS output module used.",
+                                "enum": [
+                                    "ffmpeg_hls_muxer",
+                                    "ftl_output"
```
Technically `rtmp_output` is also valid.

WizardCM commented:
@derrod Done. Also rebased on master.
